app/src/main/java/org/mozilla/fenix/settings/quicksettings/QuickSettingsFragmentStore.kt

```diff
@@ -0,0 +1,382 @@
+            val cameraPermission = PhoneFeature.CAMERA.toWebsitePermission(context, permissions, settings)
+            val microphonePermission = PhoneFeature.MICROPHONE.toWebsitePermission(context, permissions, settings)
+            val notificationPermission = PhoneFeature.NOTIFICATION.toWebsitePermission(context, permissions, settings)
+            val locationPermission = PhoneFeature.LOCATION.toWebsitePermission(context, permissions, settings)
+            val shouldBeVisible = cameraPermission.isVisible || microphonePermission.isVisible ||
+                    notificationPermission.isVisible || locationPermission.isVisible
+
+            return WebsitePermissionsState(shouldBeVisible, cameraPermission, microphonePermission,
+                notificationPermission, locationPermission
+            )
+        }
+
+        private fun PhoneFeature.toWebsitePermission(
+            context: Context,
+            permissions: SitePermissions?,
+            settings: Settings
+        ): WebsitePermission {
+            val status = getPermissionStatus(context, permissions, settings)
+            val nonexistentPermission: WebsitePermission? = null
+            return when (this) {
+                PhoneFeature.CAMERA -> WebsitePermission.Camera(
+                    status.status, status.isVisible, status.isEnabled, status.isBlockedByAndroid
+                )
+                PhoneFeature.LOCATION -> WebsitePermission.Location(
+                    status.status, status.isVisible, status.isEnabled, status.isBlockedByAndroid
+                )
+                PhoneFeature.MICROPHONE -> WebsitePermission.Microphone(
+                    status.status, status.isVisible, status.isEnabled, status.isBlockedByAndroid
+                )
+                PhoneFeature.NOTIFICATION -> WebsitePermission.Notification(
+                    status.status, status.isVisible, status.isEnabled, status.isBlockedByAndroid
+                )
+                PhoneFeature.AUTOPLAY -> nonexistentPermission!! // fail-fast
+            }
+        }
+
+        private fun PhoneFeature.getPermissionStatus(
+            context: Context,
+            permissions: SitePermissions?,
+            settings: Settings
+        ) = PermissionStatus(
+            status = getActionLabel(context, permissions, settings),
+            isVisible = shouldBeVisible(permissions, settings),
+            isEnabled = shouldBeEnabled(context, permissions, settings),
+            isBlockedByAndroid = !isAndroidPermissionGranted(context)
+        )
+
+        private data class PermissionStatus(
+            val status: String,
+            val isVisible: Boolean,
+            val isEnabled: Boolean,
+            val isBlockedByAndroid: Boolean
+        )
+    }
+}
+
+// -------------------------------------------------------------------------------------------------
+// States
+// -------------------------------------------------------------------------------------------------
```

Comment:
Optional nit: I know we generally keep these in one file, but there's enough going on here that I would personally break this file into two, or maybe even three.